src/gamepad/linux/linux-raw-gamepad-monitor.vala

```diff
@@ -0,0 +1,70 @@
+// This file is part of GNOME Games. License: GPLv3
+
+private class LibGamepad.LinuxRawGamepadMonitor : Object, RawGamepadMonitor {
+	public delegate void RawGamepadCallback (RawGamepad raw_gamepad);
+
+	private GUdev.Client client;
+	private HashTable<string, RawGamepad> raw_gamepads;
+
+	public LinuxRawGamepadMonitor () {
+		client = new GUdev.Client ({"input"});
+		client.uevent.connect (handle_udev_client_callback);
+
+		// Initialize internally plugged in gamepads
+		client.query_by_subsystem ("input").foreach (initial_device_iterator);
+	}
+
+	public void foreach_gamepad (RawGamepadCallback callback) {
+		raw_gamepads.foreach((identifier, raw_gamepad) => callback (raw_gamepad));
+	}
+
+	private void handle_udev_client_callback (string action, GUdev.Device dev) {
+		if (dev.get_device_file () == null)
+			return;
+
+		var identifier = dev.get_device_file ();
+		if ((dev.has_property ("ID_INPUT_JOYSTICK") && dev.get_property ("ID_INPUT_JOYSTICK") == "1") ||
+		    (dev.has_property (".INPUT_CLASS") && dev.get_property (".INPUT_CLASS") == "joystick")) {
```

Comment:
Let's add a `private static bool is_gamepad (GUdev.Device device)` method and use it here and in the next method.